### PR TITLE
Support multiple data source prop passwordless and new prop of credential bean name

### PR DESCRIPTION
**Pull request by moarychan**
# Description

Fixes https://github.com/Azure/azure-sdk-for-java/issues/41977

Changelogs:
- Fix primitive type prop (`isManagedIdentityEnabled`) copy issue
- Support multiple data source prop passwordless
- Add new prop of credential bean name, it's suitable for all SDK client, JMS passworless, Redis passwordless, JDBC passwordless, Spring Integration for Service Bus.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java [hunk 1]
```diff
@@ -31,6 +33,16 @@ public AzureRedisCredentials(String username, PasswordlessProperties passwordles
         this.username = resolveUsername(azureAuthenticationTemplate, username);
     }
 
+    public AzureRedisCredentials(String username,
+                                 TokenCredentialProvider tokenCredentialProvider,
+                                 PasswordlessProperties passwordlessProperties) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java:38]:
```suggestion
    public AzureRedisCredentials(String username,
                                 PasswordlessProperties passwordlessProperties,
                                 TokenCredentialProvider tokenCredentialProvider) {
```

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java [hunk 2]
```diff
@@ -31,6 +33,16 @@ public AzureRedisCredentials(String username, PasswordlessProperties passwordles
...
+        Objects.requireNonNull(passwordlessProperties, "PasswordlessProperties is required");
+        azureAuthenticationTemplate = new AzureAuthenticationTemplate(tokenCredentialProvider, null);
+        Properties properties = passwordlessProperties.toPasswordlessProperties();
+        azureAuthenticationTemplate.init(properties);
+        this.username = resolveUsername(azureAuthenticationTemplate, username);
+    }
+
     public AzureRedisCredentials(String username, PasswordlessProperties passwordlessProperties, TokenCredential tokenCredential) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java:46]: Can we call the newly added ctor in from this method?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 3]
```diff
@@ -17,28 +24,70 @@
...
+    private ServiceBusJmsConnectionFactory createJmsConnectionFactory(AzureServiceBusJmsProperties jmsProperties,
                                                                              ObjectProvider<AzureServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
-        return new ServiceBusJmsConnectionFactoryFactory(properties,
+        TokenCredentialProvider tokenCredentialProvider = getPasswordlessTokenCredentialProvider(jmsProperties);
+        return new ServiceBusJmsConnectionFactoryFactory(tokenCredentialProvider, jmsProperties,
             factoryCustomizers.orderedStream().collect(Collectors.toList()))
             .createConnectionFactory(ServiceBusJmsConnectionFactory.class);
     }
 
+    private TokenCredentialProvider getPasswordlessTokenCredentialProvider(AzureServiceBusJmsProperties serviceBusJmsProperties) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java:56]: can we leverage the authenticate template instead?

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java:56]: We should put all the same logic into the same place, instead of writing it everywhere

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/resourcemanager/AzureEventHubsResourceManagerAutoConfiguration.java [hunk 4]
```diff
@@ -42,7 +43,11 @@ public class AzureEventHubsResourceManagerAutoConfiguration extends AzureService
     @Bean
     @ConditionalOnMissingBean
     @ConditionalOnProperty(prefix = AzureEventHubsProperties.PREFIX, value = "namespace")
-    @ConditionalOnMissingProperty(prefix = AzureEventHubsProperties.PREFIX, value = "connection-string")
+    @ConditionalOnMissingProperty({
+        AzureEventHubsProperties.PREFIX + ".connection-string",
+        AzureGlobalProperties.PREFIX + ".credential.token-credential-bean-name",
+        AzureEventHubsProperties.PREFIX + ".credential.token-credential-bean-name"
+    })
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/resourcemanager/AzureEventHubsResourceManagerAutoConfiguration.java:50]: But the credential can come from all the other spring.cloud.azure.credential.xxx properties

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/servicebus/AzureServiceBusMessagingAutoConfiguration.java [hunk 5]
```diff
@@ -80,8 +84,19 @@ static class ProcessorContainerConfiguration {
         @ConditionalOnMissingBean
         ServiceBusProcessorFactory defaultServiceBusNamespaceProcessorFactory(
             NamespaceProperties properties,
-            ObjectProvider<PropertiesSupplier<ConsumerIdentifier, ProcessorProperties>> suppliers) {
-            return new DefaultServiceBusNamespaceProcessorFactory(properties, suppliers.getIfAvailable());
+            ObjectProvider<PropertiesSupplier<ConsumerIdentifier, ProcessorProperties>> suppliers,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/servicebus/AzureServiceBusMessagingAutoConfiguration.java:87]: What are these for?

Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/servicebus/AzureServiceBusMessagingAutoConfiguration.java:87]: For Service Bus Integration scenario, it's used to custom the Service Bus builder factory in `DefaultServiceBusNamespaceProcessorFactory`.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/resourcemanager/AzureEventHubsResourceManagerAutoConfiguration.java (hunk 4 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/resourcemanager/AzureEventHubsResourceManagerAutoConfiguration.java:50]: Yes, not covered all the credentials, we should add all credential types here, right?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java (hunk 3 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java:56]: Will improve the logic

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/AzureLettucePasswordlessAutoConfiguration.java [hunk 6]
```diff
@@ -51,10 +59,11 @@ AzureRedisPasswordlessProperties redisPasswordlessProperties() {
 
     @Bean(name = "azureRedisCredentials")
     @ConditionalOnMissingBean
+    @DependsOn("springTokenCredentialProviderContextProvider")
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/AzureLettucePasswordlessAutoConfiguration.java:78]: Any good suggestions?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/AzureLettucePasswordlessAutoConfiguration.java [hunk 7]
```diff
@@ -54,8 +72,12 @@ AzureRedisPasswordlessProperties redisPasswordlessProperties() {
     AzureRedisCredentials azureRedisCredentials(RedisProperties redisProperties,
                                                 AzureRedisPasswordlessProperties azureRedisPasswordlessProperties,
                                                 AzureGlobalProperties azureGlobalProperties) {
+        Properties properties = azureRedisPasswordlessProperties.toPasswordlessProperties();
+        enhancePasswordlessProperties(applicationContext, "spring.data.redis.azure",
+            azureRedisPasswordlessProperties, properties);
+        TokenCredentialProvider provider = TokenCredentialProviders.createInstance(new TokenCredentialProviderOptions(properties));
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/AzureLettucePasswordlessAutoConfiguration.java:78]: I don't think we should do it this way, we should only pass in the properties, and let the provider chosen by the azure-identity extension to resolve the bean based on the properties/options

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/resourcemanager/AzureEventHubsResourceManagerAutoConfiguration.java (hunk 4 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/resourcemanager/AzureEventHubsResourceManagerAutoConfiguration.java]: Will be improved later.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/AzureLettucePasswordlessAutoConfiguration.java [hunk 8]
```diff
@@ -54,8 +66,12 @@ AzureRedisPasswordlessProperties redisPasswordlessProperties() {
     AzureRedisCredentials azureRedisCredentials(RedisProperties redisProperties,
                                                 AzureRedisPasswordlessProperties azureRedisPasswordlessProperties,
                                                 AzureGlobalProperties azureGlobalProperties) {
-        return new AzureRedisCredentials(redisProperties.getUsername(),
-            mergeAzureProperties(azureGlobalProperties, azureRedisPasswordlessProperties));
+        AzureRedisPasswordlessProperties redisPasswordlessProperties = mergeAzureProperties(azureGlobalProperties, azureRedisPasswordlessProperties);
+        String passwordlessPropertiesPrefix = "spring.data.redis.azure";
+        registerTokenCredentialBean(applicationContext, passwordlessPropertiesPrefix, redisPasswordlessProperties);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/AzureLettucePasswordlessAutoConfiguration.java:71]: Do we really need to register the bean here, if the token credential provider can actually resolve one according to options?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java [hunk 9]
```diff
@@ -25,12 +26,20 @@ public class AzureRedisCredentials implements RedisCredentials {
...
      */
     public AzureRedisCredentials(String username, PasswordlessProperties passwordlessProperties) {
-        Objects.requireNonNull(passwordlessProperties, "PasswordlessProperties is required");
+        Objects.requireNonNull(passwordlessProperties, "PasswordlessProperties is required.");
         azureAuthenticationTemplate = new AzureAuthenticationTemplate();
         azureAuthenticationTemplate.init(passwordlessProperties.toPasswordlessProperties());
         this.username = resolveUsername(azureAuthenticationTemplate, username);
     }
 
+    public AzureRedisCredentials(String username,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java:35]: Why do we need this ctor here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jdbc/JdbcPropertiesBeanPostProcessor.java [hunk 10]
```diff
@@ -134,20 +140,13 @@ private void enhanceUserAgent(DatabaseType databaseType, JdbcConnectionStringEnh
         }
     }
 
-    private Map<String, String> buildEnhancedProperties(DatabaseType databaseType, AzureJdbcPasswordlessProperties properties) {
+    private Map<String, String> buildEnhancedProperties(String passwordlessPropertiesPrefix,
+                                                        DatabaseType databaseType,
+                                                        AzureJdbcPasswordlessProperties properties) {
+        registerTokenCredentialBean(applicationContext, passwordlessPropertiesPrefix, properties);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jdbc/JdbcPropertiesBeanPostProcessor.java:146]: Same question here, if we can construct the token credential based on the proeprties, why do we need to register the bean here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryFactory.java [hunk 11]
```diff
@@ -16,17 +16,27 @@
...
     ServiceBusJmsConnectionFactoryFactory(AzureServiceBusJmsProperties properties,
                                           List<AzureServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
         Assert.notNull(properties, "Properties must not be null");
         this.properties = properties;
         this.factoryCustomizers = (factoryCustomizers != null) ? factoryCustomizers : Collections.emptyList();
-
+        if (properties.isPasswordlessEnabled()) {
+            Properties passwordlessProperties = properties.toPasswordlessProperties();
+            enhancePasswordlessProperties(AzureServiceBusJmsProperties.PREFIX, properties, passwordlessProperties);
+            this.tokenCredentialProvider = TokenCredentialProvider.createDefault(new TokenCredentialProviderOptions(passwordlessProperties));
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryFactory.java:36]: Same here, why do we need to construct this token credential provider?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/context/AzureTokenCredentialAutoConfiguration.java [hunk 12]
```diff
@@ -101,6 +104,11 @@ AzureTokenCredentialResolver azureTokenCredentialResolver(
                 return null;
             }
 
+            String tokenCredentialBeanName = azureProperties.getCredential().getTokenCredentialBeanName();
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/context/AzureTokenCredentialAutoConfiguration.java:107]: Maybe we could reconsider this later, do we need two beans referencing the same object here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java [hunk 13]
```diff
@@ -25,9 +28,11 @@ public class AzureRedisCredentials implements RedisCredentials {
      * Create instance of Azure Redis Credentials
      */
     public AzureRedisCredentials(String username, PasswordlessProperties passwordlessProperties) {
-        Objects.requireNonNull(passwordlessProperties, "PasswordlessProperties is required");
+        Objects.requireNonNull(passwordlessProperties, "PasswordlessProperties is required.");
         azureAuthenticationTemplate = new AzureAuthenticationTemplate();
-        azureAuthenticationTemplate.init(passwordlessProperties.toPasswordlessProperties());
+        Properties properties = passwordlessProperties.toPasswordlessProperties();
+        enhancePasswordlessProperties("spring.data.redis.azure", passwordlessProperties, properties);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java:34]: What's this prefix for?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 14]
```diff
@@ -17,25 +17,33 @@
...
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.support.GenericApplicationContext;
 import org.springframework.jms.connection.CachingConnectionFactory;
 
 import java.util.stream.Collectors;
 
 @Configuration(proxyBeanMethods = false)
 @ConditionalOnMissingBean(ConnectionFactory.class)
-class ServiceBusJmsConnectionFactoryConfiguration {
+public class ServiceBusJmsConnectionFactoryConfiguration  {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java:27]: Why changing this to public?

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/factory/AbstractAzureServiceClientBuilderFactory.java [hunk 15]
```diff
@@ -343,4 +367,9 @@ public void setTokenCredentialResolver(AzureCredentialResolver<TokenCredential>
             LOGGER.debug("Will ignore the 'null' token credential resolver.");
         }
     }
+
+    @Override
+    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/factory/AbstractAzureServiceClientBuilderFactory.java:372]: Maybe later we could refactor this

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/util/AzurePasswordlessPropertiesUtils.java [hunk 16]
```diff
@@ -74,34 +70,4 @@ public static <T extends PasswordlessProperties> void mergeAzureCommonProperties
...
         copyAzureCommonPropertiesIgnoreNull(properties, target);
     }
-
-    /**
-     * Copy common properties from source object to target object. Ignore the source value if it is null.
-     *
-     * @param source The source object.
-     * @param target The target object.
-     */
-    public static void copyPropertiesIgnoreNull(Object source, Object target) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/util/AzurePasswordlessPropertiesUtils.java:84]: Why are these removed

**Review by saragluna — APPROVED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java (hunk 13 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/data/redis/lettuce/AzureRedisCredentials.java:34]: It's a debug logging format for passwordless.

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/factory/AbstractAzureServiceClientBuilderFactory.java (hunk 15 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/factory/AbstractAzureServiceClientBuilderFactory.java:372]: Yes, will continue to improve

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/util/AzurePasswordlessPropertiesUtils.java (hunk 16 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/util/AzurePasswordlessPropertiesUtils.java:84]: It's a duplicate method.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/context/AzureTokenCredentialAutoConfiguration.java (hunk 12 above)
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/context/AzureTokenCredentialAutoConfiguration.java:107]: Yes, will continue to improve
